cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,367 @@
+[[missing-null,missing `null`]]*Missing*:: A "missing" `null` is taken to be a marker for the absence of a value. In the context of updating node properties from a map, a "missing" `null` is used to mark properties that are to be removed. In the context of `DISTINCT` and grouping, a "missing" `null` value is used as grouping key for all records that miss a more specific value. Since in these cases, two "missing" `null` values represent the same concept, they should always be treated as the same value.
+
+[[regular-map]]
+==== Regular maps
+
+Cypher today has one supertype `MAP` for all map values. This includes nodes (of subtype `NODE`), relationships (of subtype `RELATIONSHIP`), and any other map (not captured by a subtype of `MAP`). For the purpose of this document, we define a regular map to be any value of type `MAP` that is neither a `NODE` nor a `RELATIONSHIP`.
+
+[[comparability-equality,comparability and equality]]
+=== Comparability and equality
+
+We propose that comparability and equality should be aligned with each other, i.e.
+
+`expr1 = expr2` if and only if `expr1 >= expr2 && expr1 \<= expr2`.
+
+Comparability and equality use <<unknown-null,"unknown" `null` values>>.
+
+[[incomparable,incomparable]]
+==== Incomparability
+
+If and only if every comparison and equality test involving a specific value evaluates to `null`, this value is said to be incomparable.
+
+Furthermore, if every comparison or equality test between two specific values evaluates to `null`, theses values are said to be incomparable with each other.
+
+[[comparability-def,comparability]]
+==== Comparability
+
+We propose that <<comparability-con,comparability>> should be defined between any pair of values, as specified below.
+
+- General rules
+  * Values are only comparable within their most specific type (except for numbers, see below).
+  * Equal values are grouped together.
+- Numbers
+  * Integers are compared numerically in ascending order.
+  * Floats (excluding `NaN` and the Infinities) are compared numerically in ascending order.
+  * Numbers of different types (excluding `NaN` and the Infinities) are compared to each other as if both numbers would have been coerced to arbitrary precision big decimals (currently outside the Cypher type system) before comparing them with each other numerically in ascending order.
+  * Positive infinity is of type `FLOAT`, equal to itself and larger than any other number (excluding `NaN`).
+  * Negative infinity is of type `FLOAT`, equal to itself and smaller than any other number (excluding `NaN`).
+  * `NaN` is <<incomparable>>.
+  * Numbers are <<incomparable>> to any value that is not also a number.
+- Booleans
+  * Booleans are compared such that `false` is less than `true`.
```

Comment:
I agree that we want this for _orderability_, but I maintain my position that values could be orderable without being comparable. I.e. you don't necessarily have to be able to apply the `<` and `>` operators to the values in order to be able to `ORDER BY` them. I think it would be sensible for orderability to be "broader" than comparability, as long as it agrees with comparability in all cases where comparability applies.